OAuth docs (#334)
r-lib/httr2: vignettes/articles/oauth.Rmd

library(httr2)
```

OAuth[^1] is an authorization framework design for performing work on the behalf of a user.

> You've probably used it a bunch without knowing what it's called: it's used when you login to a non-Google website using your Google account, when you give an phone access to your twitter account, or when you login to a streaming app on your smart TV.

Reviewer: ```suggestion
You've probably used it a bunch without knowing what it's called: it's used when you login to a non-Google website using your Google account, when you give a phone access to your twitter account, or when you login to a streaming app on your smart TV.
```

[^1]: Here I'll only talk about OAuth 2.0 which is the only version in common use today.
    OAuth 1.0 is largely only of historical interest.

You'll notice none of these common scenarios quite match to using an R package to scrape data from a web API, and that's the pleasure and pain of OAuth.
OAuth gives you the incredible power to extract data from popular web services, but it's fundamentally designed for a different use case.
This means that while httr2 can make the interactive experience pretty seamless, unattended usage (i.e. on CI or in an automated script) is usually going to include some pain.

> This vignette builds on the techniques in discussed in `vignette("wrapping-apis")`, so I recommend you read that vignette first.

Reviewer: ```suggestion
This vignette builds on the techniques discussed in `vignette("wrapping-apis")`, so I recommend you read that vignette first.
```

> I'll also assume that you're working in package, since most people want to wrap up an API into a bunch of R functions, but most of the ideas will also apply if you're creating a one off script.

Reviewer: ```suggestion
I'll also assume that you're working in a package, since most people want to wrap up an API into a bunch of R functions, but most of the ideas will also apply if you're creating a one off script.
```

## OAuth basics

> OAuth is a broad framework that has many different variants, called **flows**, which makes it hard to provide sweeping generalisations, but the basic idea of OAuth is to create a hierarchy of increasing more specific and shorter lived credentials, so that the impact of a credential being lost is as small as possible.

Reviewer: ```suggestion
OAuth is a broad framework that has many different variants, called **flows**, which makes it hard to provide sweeping generalisations, but the basic idea of OAuth is to create a hierarchy of increasingly more specific and shorter-lived credentials, so that the impact of a credential being lost is as small as possible.
```

```

OAuth[^1] is a authorization framework design for performing work on the behalf of a user.
It allows you to wrap an API that requires a user login without your app ever directly seeing the user's user name and password.
You've almost certainly used this before on the web or with your TV; it's used in most cases where one website (or your TV) needs to login to an account on your behalf.

[^1]: Here I'll only talk about OAuth 2.0 which is the only version in common use today.
    OAuth 1.0 is largely only of historical interest.

OAuth is a broad framework that has many different variants, called **flows**, which makes it hard to provide generalisable advice.
The one commonality is that all flows eventually generate an **access token** (essentially a random string of numbers and letters) which is provided to the API in the form of a **bearer token** in the `Authorization` header.
If you have this access token, you can use it directly with `req_auth_bearer_token()`, but in most cases you will want to give httr2 the information to generate and use the access token for you by calling a `req_oauth_` function.
The rest of the vignette describes just how to do that.

## Clients

> The first step in working with any OAuth API is to create an **application** **client**, which I'll generally just call a **client**.

Reviewer: Maybe something like, "In this context, the script or R package you are writing is the "application."

You'll do this by registering for a developer account on the API's website then creating a new OAuth app.
This process varies from API to API (it's normal to spend some time hunting through the docs and settings), but at the end of it you'll get a client id (another random string of numbers and letters).

Sometimes the client id is all you need, and you can give httr2 the information about the client with a call to `oauth_client()`, something like this:

```{r}
client <- oauth_client(
  id = "28acfec0674bb3da9f38",
  token_url = "https://github.com/login/oauth/access_token",
  name = "hadley-oauth-test-2"
)
```

The call to `oauth_client()` also includes a `name` and a `token_url`.
The `name` is human facing and should typically be the same as the package, if you're using one.
I have a bunch of apps that I've used for testing, so here I've and is used to remind you which app the client corresponds to.
The `token_url` points to the URL that's used to obtain an access token.

> You'll need to find this from the documentation for the API you're wrapping; it will typically be found in the section that describes the OAuth process and will be an endpoint that returns an access token.

Author: I think of endpoint as meaning URL + HTTP method, but I'll think of some way of rewording this to make it more clear that the auth endpoints often feel bolted on.

### Client secret

In most cases, however, the API will also require a client secret.
While this is called a secret, it's typically not that important to keep it a secret because of two reasons:

-   It's typically easy to create an new app on the developer website so stealing yours wouldn't save much time.

> -   It's unusual for an OAuth client to be able to do anything in its own right, so stealing your secret doesn't have much benefit.

Author: Yeah but there’s no reason they can’t create their own app that has the same details as your app.

> -   It's unusual for an OAuth client to be able to do anything in its own right, so stealing your secret doesn't have much benefit.

Reviewer: Hmm. Good point.

> -   It's unusual for an OAuth client to be able to do anything in its own right, so stealing your secret doesn't have much benefit.

Author: I might add a sentence saying something like "as long as you didn't pay to get the app, or have to supply any secret information when creating it, there's no reason to worry about it"

> -   It's unusual for an OAuth client to be able to do anything in its own right, so stealing your secret doesn't have much benefit.

Reviewer 2: I also feel like the word "secret" is a huge reason people assume that this needs to be kept secret in the same way that a password does. And it is a reasonable assumption!

> -   It's unusual for an OAuth client to be able to do anything in its own right, so stealing your secret doesn't have much benefit.

Reviewer: The case I can think of where the secret matters is something like a Slack app, where the app (which can be pretty much just an API client) gets approved for a given Slack workspace, and thus you can do a lot if you successfully get someone to log into the app. That's assuaged online by restricting redirect urls, but it's messy if you're hitting the API from your machine (where the redirect is going to be localhost).

That's a case where it might be best to not supply the client, although I'm still unsure how to feel about it. I guess a way to protect it is to make it clear what the app is supposed to be paired with (if you're able to do so on the approval page), but it's an area that always feels scary.

You've probably used it a bunch without knowing what it's called: it's used when you login to a non-Google website using your Google account, when you give an phone access to your twitter account, or when you login to a streaming app on your smart TV.

[^1]: Here I'll only talk about OAuth 2.0 which is the only version in common use today.
    OAuth 1.0 is largely only of historical interest.

You'll notice none of these common scenarios quite match to using an R package to scrape data from a web API, and that's the pleasure and pain of OAuth.
OAuth gives you the incredible power to extract data from popular web services, but it's fundamentally designed for a different use case.
This means that while httr2 can make the interactive experience pretty seamless, unattended usage (i.e. on CI or in an automated script) is usually going to include some pain.

This vignette builds on the techniques in discussed in `vignette("wrapping-apis")`, so I recommend you read that vignette first.
I'll also assume that you're working in package, since most people want to wrap up an API into a bunch of R functions, but most of the ideas will also apply if you're creating a one off script.

> ## OAuth basics

Author: @llrs @maelle @jonthegeek thanks for all your feedback so far 😄 I added this new section and if you have time to take a look, I'd definitely appreciate your thoughts!

OAuth gives you the incredible power to extract data from popular web services, but it's fundamentally designed for a different use case.
This means that while httr2 can make the interactive experience pretty seamless, unattended usage (i.e. on CI or in an automated script) is usually going to include some pain.

This vignette builds on the techniques in discussed in `vignette("wrapping-apis")`, so I recommend you read that vignette first.
I'll also assume that you're working in package, since most people want to wrap up an API into a bunch of R functions, but most of the ideas will also apply if you're creating a one off script.

## OAuth basics

OAuth is a broad framework that has many different variants, called **flows**, which makes it hard to provide sweeping generalisations, but the basic idea of OAuth is to create a hierarchy of increasing more specific and shorter lived credentials, so that the impact of a credential being lost is as small as possible.

The longest lived and most powerful credential is typically a user name and password.

> Most people don't change their passwords regularly, and often (against all advice) reuse the same password on multiple websites.

Reviewer 3: would this be the occasion to preach for password managers by simply linking to some sort of password management 101? I don't have such a link at hand though. :innocent: (in the R-Ladies guide there's https://guide.rladies.org/organization/tech/security/#use-a-personal-password-manager)

And if you have a user name and password, you have total control over that account; you can even use it to change the password so the actual user can't log in.
That means as a programmer you never want to touch user name-password pairs because if they're lost or stolen, they give very wide access.

Avoiding that problem lead to the creation of OAuth.
The basic idea is that instead of your package asking the user to give you their user name and password, you instead ask them to log in and give your package permission to use the API on their behalf.
The API gives this permission in the form of an **access token**, which is essentially a random string of numbers and letters.

> An access token tends to be very short lived, lasting maybe only days or hours, and is bound to a specified **scope** of access.

Reviewer 3: add examples of scope? for instance `(read user info, write, delete)`

The access token has some big advantages over a user name and password:

-   It's short-lived so if it's lost or stolen, there's only a limited amount of time where it can be abused.
-   It has limited scope, so even if stolen, it can't be used to do something particularly nefarious like changing your password or contact details.
-   It's bound to a specific application, so it can be invalidated (cancelled) without affecting any other uses.

> If you have an access token you can use it to authenticate an API by passing a **bearer token** in the `Authorization` header, which you can do with `req_auth_bearer_token()`.

Reviewer 3: isn't the bearer the access token then? it's a bit unclear here

Author: Changed this to "If you have an access token you can use it to authenticate an API by passing it as a bearer token in the Authorization header". Hopefully that's more clear.

Reviewer 3: it is!!

However, in most cases you will want to let httr2 manage this by calling one of the `req_oauth_` functions we'll talk about shortly.

One of the reasons that you want httr2 to manage your tokens is that because access tokens are so short lived, they're often accompanied by a **refresh token**.

> A refresh token lasts for a longer amount of time and has one job: it allows you to get a new access token when the previous one expires.

Reviewer 3: to me that's always been confusing because losing the refresh token seems dangerous so I'd appreciate some comment on that.

Author: Don’t lose it 😀

This leads to a hierarchy of credentials from strongest to weakest:

-   user name + password: your code should never touch these!
-   refresh token: it's okay to store this on your computer.
-   access token: you submit this with every request.

Now that you've got the basic idea of OAuth, lets talk about the details.

## Clients

The first step in working with any OAuth API is to create an **application** **client**.
It's called an application client because in the wider world OAuth is usually used with a web, phone, or tv app, but here your "app" is going to be your R package.
For this reason, httr2 calls the application client a **client**, but many APIs will call it an OAuth application.

To create a client, you'll need to first register for a developer account on the API's website.
In most cases this is easy, totally automated, and only takes a couple of minutes.

> That will give you access to some sort of developer portal which you can then use to create a new OAuth app (aka a client).

Reviewer 3: to create -> to register (create sounds too much like actually building the thing)

An access token is very short lived, lasting maybe only days, and is bound to a specified **scope** of access.
The access token has some big advantages over a user name and password:

-   It's short-lived so if it's lost or stolen, there's only a limited amount of time where it can be abused.
-   It has limited scope, so even if stolen, it can't be used to do something particularly nefarious like changing your password or contact details.
-   It's bound to a specific application, so it can be invalidated (cancelled) without affecting any other uses.

If you have an access token you can use it to authenticate an API by passing it as a **bearer token** in the `Authorization` header, which you can do with `req_auth_bearer_token()`.
However, in most cases you will want to let httr2 manage this by calling one of the `req_oauth_` functions we'll talk about shortly.

One of the reasons that you want httr2 to manage your tokens is that because access tokens are so short lived, they're often accompanied by a **refresh token**.
A refresh token lasts for a longer amount of time and has one job: it allows you to get a new access token when the previous one expires.

> You need to look after an refresh token a little more carefully than an access token.

Reviewer: ```suggestion
You need to look after a refresh token a little more carefully than an access token.
```

In particular, you should never include a refresh token in a HTTP request as that's the job of the access token.

Overall this leads to a hierarchy of credentials from weakest to strongest:

> -   The access token only lasts a couple hours and needs to be submitted this with every request.

Reviewer: ```suggestion
-   The access token usually only lasts a couple hours and needs to be submitted with every request.
```

Some APIs are bad about the hierarchy and essentially treat access tokens like refresh tokens.

client
```

You'll probably want to wrap this up in a function in your package.
For an example, you might take a look at `example_github_client()`, which I provide in httr2 because I use it in a couple of different OAuth examples.

## Authorization code flow

Once you have a client you need to use it with a **flow** in order to get a token.
OAuth provides a number of different "flows", the most common is the "authorisation code" flow, which is implemented by a pair of functions `oauth_flow_auth_code()` and `req_oauth_auth_code()`.

I'd recommend starting with the `oauth_flow_` function.

> This isn't going to be useful in the long run, but it's the fastest way to iterate to ensure that you can get an OAuth token interactively.

Author: I've hopefully made this more clear now.

)
```

You can certainly uncover my client secret if you are an experienced R programmer and are willing to spend a bit of time experimenting, but I'm pretty sure it'd be easy for you to just create your own app on GitHub.

### Packaging

> I recommend wrapping client creation in a function in your package, e.g.:

Reviewer 3: Should there be a sentence at the beginning explaining you assume most docs readers will be packaging their OAuth code but that the vignette still applies if you're writing a script?

  ) %>% 
  req_perform() %>% 
  resp_body_json() %>% 
  str()
```

When you run this code, you'll see something like this, but it will obviously contain information about you, not me.

```{r}
#| eval: false
#> List of 32
#>  $ login        : chr "hadley"
#>  $ id           : int 4196
#>  $ node_id      : chr "MDQ6VXNlcjQxOTY="
#>  $ avatar_url   : chr "https://avatars.githubusercontent.com/u/4196?v=4"
#>  $ gravatar_id  : chr ""
#>  $ url          : chr "https://api.github.com/users/hadley"
#>  $ html_url     : chr "https://github.com/hadley"
#>  ...
#>  $ type         : chr "User"
#>  $ site_admin   : logi FALSE
#>  $ name         : chr "Hadley Wickham"
#>  $ company      : chr "@posit-pbc"
#>  $ blog         : chr "http://hadley.nz"
#>  $ location     : chr "Houston, TX"
```

### Token management

There are two big reasons to allow httr2 to manage tokens for you.
The first is that httr2 will automatically refresh the token if it's expired.
The second is cross-session caching which you can turn on with `cache_disk = TRUE`.
Turning it on means that your users will only need to re-authenticate once a month, but does means their credentials are saved on disk.

> Reading the "Caching" section in `?req_oauth_auth_code` for more details.

Reviewer 3: why not reproduce it here? There are details on the not secret secret at the beginning of the vignette so why not also have the details on the actual secrets here? :smile_cat:

Author: Mostly because I don’t want to have the same information in two places because I’ll forgot to update one. Do you think it would make more sense for the vignette to be the primary source and I just link to it from the docs?

Reviewer 3: Yes, it'd be nice for the vignette to have all the info (for Ctrl+F, and for a reader to know they'll have all the info by the time they read it)

    It's useful for testing.

httr2 doesn't support the implicit grant flow.
This was historically important but is now [mostly deprecated](https://developer.okta.com/blog/2019/05/01/is-the-oauth-implicit-flow-dead) and was never a particularly good fit for R because it relies on a technique for returning the access token that only reliably works inside a web browser.

Regardless of which flow you use, you'll need to follow the same process as the example above: first get figure out how to get a token using the `oauth_flow_` function, then actually use with a request, by calling the matching `req_oauth_` function.

One additional wrinkle is that many APIs don't implement the flow in exactly the same way as the spec so httr2's built-in flows might not work at all.
If your initial attempt doesn't work, you're going to need to do some sleuthing.
This is going to be painful, but unfortunately there's no way around it.
I recommend using `with_verbosity()` so you can see exactly what httr2 is sending to the server.
You'll then need to carefully compare this to the API documentation and play "spot the difference".

> You're very welcome to file an issue and I'll do my best to help you out.

Author: Less intimidating, but I don't read them so also less helpful 😆

    This is a general problem with OAuth docs; often they use different terms for the same things.
    In httr2, I've tried to stick as closely as possible to the terms used in the official RFC specifications.

1.  httr2 opens a browser using the **authorization URL** provided by the API. The URL includes parameters that identify your app and what **scope** of access you're looking for (e.g. `tweet.read`, `userinfo.write`).
2.  The user logs in using their user name and password (hopefully using a [password manager](https://www.nytimes.com/wirecutter/blog/why-you-need-a-password-manager-yes-you/)) and approves the request.
3.  The API sends an **authorization code** back to httr2 using a callback URL that was supplied in the initial request.
4.  httr2 sends the authorization code to the **token URL** to get an access token.

In httr2 this flow is implemented by a pair of functions: `oauth_flow_auth_code()` and `req_oauth_auth_code()`.
Start with `oauth_flow_auth_code()` to check that you have all the parameters correctly specified, then use `req_oauth_auth_code()` to authenticate requests.
These two steps are described in the following sections.

### Creating a token

> `oauth_flow_auth_code()` is the best way to verify you've correctly specified the parameters to the client and the `auth_url`, without depending on correclty understanding any other part of the API.

Reviewer 2: ```suggestion
`oauth_flow_auth_code()` is the best way to verify you've correctly specified the parameters to the client and the `auth_url`, without depending on correctly understanding any other part of the API.
```

You'll need this in order to run tests for your package, but you'll probably also want to use it as the default client for your users.
In some cases, it will be necessary for each user to create their own app and matching client (e.g. if rate limits are applied by app, not by user), but this is much less user friendly so you should avoid it if possible.
That said, you should always provide a way for a user to supply their own client when you bundle a default.
You can see an example of this in [the googledrive package](https://googledrive.tidyverse.org/articles/bring-your-own-app.html).

## Authorization code flow

Once you have a client you need to use it with a flow in order to get a token.
You'll need to read the docs for your API to figure out which flows it supports, but the most common is the **authorization code**[^2] flow, which works something like this:

[^2]: Confusingly, Github calls this the "web application flow", but this is not the official name.
    This is a general problem with OAuth docs; often they use different terms for the same things.
    In httr2, I've tried to stick as closely as possible to the terms used in the official RFC specifications.

> 1.  httr2 sends opens a browser using the **authorization URL** provided by the API. The URL includes parameters that identify your app and what **scope** of access you're looking for.

Reviewer: ```suggestion
1.  httr2 opens a browser using the **authorization URL** provided by the API. The URL includes parameters that identify your app and what **scope** of access you're looking for.
```

    This is a general problem with OAuth docs; often they use different terms for the same things.
    In httr2, I've tried to stick as closely as possible to the terms used in the official RFC specifications.

1.  httr2 sends opens a browser using the **authorization URL** provided by the API. The URL includes parameters that identify your app and what **scope** of access you're looking for.

> 2.  The user logs in using their user name and password (hopefully in a password manageer) and approves the request.

Reviewer 3: maybe put a password manager link _here_ then :grin:

```

### Caching

There are two big reasons to allow httr2 to manage tokens for you.
The first is that httr2 will automatically refresh the token if it's expired.
The second is cross-session caching, as described below.

By default, the OAuth token will be cached in memory.
That means that you will only need to authenticate once in the current session, but you'll need to re-authenticate if you restart R.
In some cases, you may want to save the tokens (refresh and access) so that they can be used across sessions.
This is easy to do (just set `cache_disk = TRUE`) but you need to think through the consequences of saving the refresh token on disk.

httr2 does the best it can to save these credentials securely.
They are stored in a local cache directory (`oauth_cache_path()`) that is accessible to the current user, and they are encrypted so they will be hard for any package other than httr2 to read.
However, there's no way to prevent other R code from using httr2 to access them, so if you do choose to cache tokens, you should inform the user and give them the ability to opt-out.
httr2 automatically deletes any cached tokens that are older than 30 days whenever it's loaded.

> This means that you'll need to re-auth at least once a month, but prevents tokens for hanging around on disk long after you've forgotten you created them.

Reviewer: ```suggestion
This means that you'll need to re-auth at least once a month, but prevents tokens from hanging around on disk long after you've forgotten you created them.
```

You can see which clients have cached tokens by looking in the cache directory used by httr2:

```{r}
#| eval: false
dir(oauth_cache_path(), recursive = TRUE)
```

> Each client gets its own subdirectory named using the client `name`, so if you turn caching on, it's particularly important to give you client a good name so that the user can easily tell with package the tokens belong to.

Reviewer: ```suggestion
Each client gets its own subdirectory named using the client `name`, so if you turn caching on, it's particularly important to give your client a good name so that the user can easily tell which package the tokens belong to.
```

#>  $ location     : chr "Houston, TX"
```

> ### Caching

Reviewer 3: In the intro to OAuth session you had written "refresh token: it's okay to store this on your computer" but here you mention the access and refresh tokens will be saved to disk so I'd recommend updating the intro to OAuth session to minimize confusion.

Reviewer 3: especially as you write "you need to think through the consequences of saving the refresh token on disk."
